index.js

```diff
@@ -257,6 +258,49 @@ class KEAFacility extends BaseFacility {
     return ip
   }
 
+  async _assignIp ({ mac, ip, subnetId }) {
+    debug('assignIp', mac, ip, subnetId)
+    if (!mac || !ip || !subnetId) {
+      throw new Error('ERR_MAC_AND_IP_AND_SUBNETID_REQUIRED')
+    }
+    const subnet = this.subnets.find(s => s.id === subnetId)
+    if (!subnet) {
+      throw new Error('ERR_SUBNET_NOT_FOUND')
+    }
+    const lease = this.leases.find(l => l.mac.toLowerCase() === mac.toLowerCase())
+    if (lease) {
+      if (lease.ip !== ip) {
+        throw new Error('ERR_IP_NOT_MATCH')
+      }
+      if (lease.subnetId !== subnetId) {
+        throw new Error('ERR_IN_ANOTHER_SUBNET')
+      }
+      return ip
+    }
+    await this.setLeases([{
+      ip,
+      mac,
+      subnetId
+    }])
+    return ip
+  }
+
+  async _assignIps (reqs) {
+    await this._prepareLeases()
+    const res = []
+    for (let i = 0; i < reqs.length; i++) {
+      const req = reqs[i]
+      res.push(this._addJob({
+        payload: { mac: req.mac, ip: req.ip, subnetId: req.subnetId },
+        respKey: 'ip',
+        process: async () => {
+          return await this._assignIp(req, true)
+        }
+      }))
+    }
+    return (await Promise.allSettled(res)).map(r => ({ success: r.value.success, mac: r.value.request.mac, ip: r.value.response, error: r.value.error }))
```

Comment:
why new functions, didn't we already have setIP, setIPs

Reply:
setIps, and setIps, take a subnet, and then find an empty ip in the subnet and use the setLease method, here the ip is coming from the request